flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/StopCondition.java

```diff
@@ -0,0 +1,129 @@
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.connector.pulsar.source;
+
+import org.apache.pulsar.client.api.Consumer;
+import org.apache.pulsar.client.api.Message;
+import org.apache.pulsar.client.api.MessageId;
+import org.apache.pulsar.client.api.PulsarClientException;
+import org.apache.pulsar.client.impl.MessageIdImpl;
+
+import java.io.Serializable;
+import java.util.Comparator;
+import java.util.Map;
+
+import static org.apache.flink.connector.pulsar.source.StopCondition.StopResult.DONT_STOP;
+import static org.apache.flink.connector.pulsar.source.StopCondition.StopResult.STOP_AFTER;
+import static org.apache.flink.connector.pulsar.source.StopCondition.StopResult.STOP_BEFORE;
+
+/** An interface to control when to stop. */
+public interface StopCondition extends Serializable {
+    StopResult shouldStop(AbstractPartition partition, Message<?> message);
+
+    /** Enum for stop condition. */
+    enum StopResult {
+        STOP_BEFORE,
+        STOP_AFTER,
+        DONT_STOP;
+    }
+
+    Comparator<MessageId> NON_BATCH_COMPARATOR =
+            new Comparator<MessageId>() {
+                final Comparator<MessageIdImpl> implComparator =
+                        Comparator.comparingLong(MessageIdImpl::getLedgerId)
+                                .thenComparingLong(MessageIdImpl::getEntryId)
+                                .thenComparingInt(MessageIdImpl::getPartitionIndex);
+
+                @Override
+                public int compare(MessageId o1, MessageId o2) {
+                    return implComparator.compare((MessageIdImpl) o1, (MessageIdImpl) o2);
+                }
+            };
+
+    default void init(AbstractPartition partition, Consumer<byte[]> consumer)
+            throws PulsarClientException {}
+
+    static StopCondition stopAtMessageId(MessageId id) {
+        return (partition, message) -> hitMessageId(message, id) ? STOP_BEFORE : DONT_STOP;
+    }
+
+    static boolean hitMessageId(Message<?> message, MessageId id) {
+        return NON_BATCH_COMPARATOR.compare(message.getMessageId(), id) >= 0;
+    }
+
+    static StopCondition stopAfterMessageId(MessageId id) {
+        return (partition, message) -> hitMessageId(message, id) ? STOP_AFTER : DONT_STOP;
```

Comment:
This is probably not correct: if a partition misses the message id, it would read one more message.

I guess `hitMessageId` should return the comparator value such that you can distinguish between `>= 0` and `> 0`.

Reply:
I guess in the same way, it's actually possible to get rid of STOP_AFTER completely and just return a boolean.

Reply:
I think you are right, I would remove this enum class.